src/library/compiler/ir_interpreter.cpp

```diff
@@ -0,0 +1,759 @@
+                       tout() << "\n";);
+       });
+    }
+
+    /** \brief Return cached lookup result for given unmangled function name in the current binary. */
+    symbol_cache_entry lookup_symbol(name const & fn) {
+        if (symbol_cache_entry const * e = m_symbol_cache.find(fn)) {
+            return *e;
+        } else {
+            string_ref mangled = name_mangle(fn, *g_mangle_prefix);
+            string_ref boxed_mangled(string_append(mangled.to_obj_arg(), g_boxed_mangled_suffix->raw()));
+            symbol_cache_entry e_new;
+            // check for boxed version first
+            if (void *p_boxed = lookup_symbol_in_cur_exe(boxed_mangled.data())) {
+                e_new = symbol_cache_entry { p_boxed, true };
+            } else if (void *p = lookup_symbol_in_cur_exe(mangled.data())) {
+                // if there is no boxed version, there are no unboxed parameters, so use default version
+                e_new = symbol_cache_entry { p, false };
+            } else {
+                e_new = symbol_cache_entry { nullptr, false };
+            }
+            m_symbol_cache.insert(fn, e_new);
+            return e_new;
+        }
+    }
+
+    /** \brief Retrieve Lean declaration from environment. */
+    decl get_decl(name const & fn) {
+        option_ref<decl> d = find_env_decl(m_env, fn);
+        if (!d) {
+            throw exception(sstream() << "unknown declaration '" << fn << "'");
+        }
+        return d.get().value();
+    }
+
+    /** \brief Retrieve Lean definition from environment. */
+    decl get_fdecl(name const & fn) {
+        decl d = get_decl(fn);
+        if (decl_tag(d) == decl_kind::Extern) {
+            throw exception(sstream() << "unexpected external declaration '" << fn << "'");
+        }
+        return d;
+    }
+
+    /** \brief Evaluate nullary function ("constant"). */
+    object * load(name const & fn, type t) {
+        if (void * p = lookup_symbol(fn).m_addr) {
+            // constants do not have boxed wrappers, but we'll survive
+            switch (t) {
+                case type::Float: throw exception("floats are not supported yet");
+                case type::UInt8: return box(*static_cast<uint8 *>(p));
+                case type::UInt16: return box(*static_cast<uint16 *>(p));
+                case type::UInt32: return box_uint32(*static_cast<uint32 *>(p));
+                case type::UInt64: return box_uint64(*static_cast<uint64 *>(p));
+                case type::USize: return box_size_t(*static_cast<size_t *>(p));
+                case type::Object:
+                case type::TObject:
+                    return *static_cast<object **>(p);
+                default:
+                    throw exception("invalid type");
+            }
+        } else {
+            push_frame(fn, m_arg_stack.size());
+            decl d = get_fdecl(fn);
+            object * r = eval_body(decl_fun_body(d));
+            pop_frame(r);
+            return r;
```

Comment:
It seems the interpreter does not cache the value of constants.

Reply:
It was in `eval_expr`, moved